v2/yarpcgrpcfx/yarpcgrpcfx.go

```diff
@@ -0,0 +1,259 @@
+type StartInboundsParams struct {
+	fx.In
+
+	Lifecycle fx.Lifecycle
+	Router    yarpc.Router
+	Config    InboundConfig
+	Logger    *zap.Logger        `optional:"true"`
+	Tracer    opentracing.Tracer `optional:"true"`
+}
+
+// StartInbounds constructs and starts inbounds.
+func StartInbounds(p StartInboundsParams) error {
+	inbound := yarpcgrpc.Inbound{
+		Addr:   p.Config.Address,
+		Router: p.Router,
+		Tracer: p.Tracer,
+		Logger: p.Logger,
+	}
+	p.Lifecycle.Append(fx.Hook{
+		OnStart: func(ctx context.Context) error {
+			return inbound.Start(ctx)
+		},
+		OnStop: func(ctx context.Context) error {
+			return inbound.Stop(ctx)
+		},
+	})
+	return nil
+}
+
+// OutboundsConfig is the configuration for constructing a set of outbounds.
+type OutboundsConfig struct {
+	Outbounds map[string]OutboundConfig `yaml:",inline"`
+}
+
+// OutboundConfig is the configuration for constructing a specific outbound.
+type OutboundConfig struct {
+	// Specifies the outbound's service name.
+	//
+	// If not set, defaults to the configured outbound key.
+	Service string `yaml:"name"`
+
+	// Specifies the address to use for this Outbound.
+	Address string `yaml:"address"`
+
+	// Specifies the peer list chooser to use for this Outbound.
+	//
+	// If set, an address does not need to be configured.
+	Chooser string `yaml:"chooser"`
+}
+
+// OutboundsConfigParams defines the dependencies of this module.
+type OutboundsConfigParams struct {
+	fx.In
+
+	Provider config.Provider
+}
+
+// OutboundsConfigResult defines the values produced by this module.
+type OutboundsConfigResult struct {
+	fx.Out
+
+	Config OutboundsConfig
+}
+
+// NewOutboundsConfig produces an OutboundsConfig.
+func NewOutboundsConfig(p OutboundsConfigParams) (OutboundsConfigResult, error) {
+	oc := OutboundsConfig{}
+	if err := p.Provider.Get(_outboundConfigurationKey).Populate(&oc); err != nil {
+		return OutboundsConfigResult{}, err
+	}
+	return OutboundsConfigResult{
+		Config: oc,
+	}, nil
+}
+
+// ClientParams defines the dependencies of this module.
+type ClientParams struct {
+	fx.In
+
+	Lifecycle       fx.Lifecycle
+	Config          OutboundsConfig
+	Dialer          *yarpcgrpc.Dialer
+	ChooserProvider yarpc.ChooserProvider
+
+	Logger *zap.Logger        `optional:"true"`
+	Tracer opentracing.Tracer `optional:"true"`
+}
+
+// ClientResult defines the values produced by this module.
+type ClientResult struct {
+	fx.Out
+
+	Clients []yarpc.Client `group:"yarpcfx"`
+}
+
+// NewClients produces yarpc.Clients.
+func NewClients(p ClientParams) (ClientResult, error) {
+	var clients []yarpc.Client
+	for name, o := range p.Config.Outbounds {
+		var (
+			chooser yarpc.Chooser
+			addr    *url.URL
+		)
+		if o.Chooser != "" {
+			var ok bool
+			chooser, ok = p.ChooserProvider.Chooser(o.Chooser)
+			if !ok {
+				return ClientResult{}, fmt.Errorf("failed to resolve outbound peer list chooser: %q", o.Chooser)
+			}
+		} else {
+			addr = &url.URL{Host: o.Address}
```

Comment:
If we can avoid the error check altogether from `url.Parse`, what are your thoughts on always applying the address rather than in the `else` branch? If a chooser is configured, it will use it instead anyways, right?

With that said, we're now deferring the validation on the configured `Address` until a later point. It might be better to explicitly validate the error here rather than later. 

What are your thoughts? I'm open to either approach as long as the above cases are accounted for. Either way, let's make sure that `yarpchttpfx` is consistent with what we do here.

Reply:
yeah that's fair, I'll update this to the parse it like we do in `yarpchttpfx` to fail early. `Parse` fails when no scheme is given, so we'll probably have to revisit anyways. Otherwise, users will have to to add `http://` to their gRPC address, which may be confusing

Reply:
Ah yea that's right.. we'll definitely want to avoid enforcing the scheme in the configuration.

Reply:
yeah, after playing around, i'm actually more inclined to leave it as is to avoid this. The dialer can accept an address with and without `http://` if we leave it like this.

Reply:
Cool that's fine with me. We can always revisit it if need be.